chroot-bin/croutonnotify

```diff
@@ -0,0 +1,83 @@
+#!/bin/sh -e
+# Copyright (c) 2015 The crouton Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+USAGE="${0##*/} -j
+${0##*/} -t title [-m message] [-d display] [-i id]
+Raises a notification in Chromium OS (requires crouton extension).
+
+When -j is specified, reads JSON data from stdin, in the format required
+by chrome.notifications API, with 2 additional fields, \"crouton_id\" and
+\"crouton_display\", corresponding respectively to the notification id, and the
+display to switch to when the notification is clicked (croutoncycle parameter).
+
+Otherwise, constructs a \"basic\" notification with the requested fields.
+
+Options:
+    -j       Read JSON data from stdin (see example below)
+    -t       Title to display (chrome.notifications field: \"title\")
+    -m       Message to display (chrome.notifications field: \"message\")
+    -d       Display to switch to when the notification is clicked (croutoncycle
+             parameter). Default: Do not switch display.
+    -i       ID to pass to chrome.notifications.create: only the last
+             notification with a given ID is shown.
+             Default: Display a new notification.
+
+Example JSON data:
+"'{
+  "type": "basic",
+  "title": "Primary Title",
+  "message": "Primary message to display",
+  "crouton_display": ":1"
+}'
+
+. "$(dirname "$0")/../installer/functions"
+
+DISPLAY=""
+MESSAGE=""
+TITLE=""
+ID=""
+JSON=""
+
+# Process arguments
+while getopts 'd:i:jm:t:' f; do
+    case "$f" in
+    d) DISPLAY="$OPTARG";;
+    i) ID="$OPTARG";;
+    j) JSON="y";;
+    m) MESSAGE="$OPTARG";;
+    t) TITLE="$OPTARG";;
+    \?) error 2 "$USAGE";;
+    esac
+done
+
+# No extra parameters, -j precludes other parameters, and at least title
+# must be specified (or -j)
+if [ "$#" != "$((OPTIND-1))" ] ||
+        [ "$JSON" = 'y' -a -n "$DISPLAY$ID$MESSAGE$TITLE" ] ||
+        [ -z "$JSON$TITLE" ]; then
+    error 2 "$USAGE"
+fi
+
+JSONDATA=""
+if [ -z "$JSON" ]; then
+    JSONDATA='{
+        "type": "'"basic"'",
+        "title": "'"$TITLE"'",
```

Comment:
TITLE, MESSAGE, DISPLAY, and ID parameters [should all be escaped for JSON](http://stackoverflow.com/questions/3020094/how-should-i-escape-strings-in-json) to avoid nasty surprises.

dash and bash are inconsistent about processing escape sequences when expanding variables, but it looks like they both leave \uXXXX sequences intact.  So avoid converting \ into \\'s and use the \uXXXX code instead, and it should remain intact as you use the variables.